extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -493,8 +556,176 @@ public void stopGracefully()
     };
   }
 
+  @GET
+  @Path("/offsets/current")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Map<Integer, Long> getCurrentOffsets()
+  {
+    return nextOffsets;
+  }
+
+  @GET
+  @Path("/offsets/end")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Map<Integer, Long> getEndOffsets()
+  {
+    return endOffsets;
+  }
+
+  @POST
+  @Path("/offsets/end")
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response setEndOffsets(
+      Map<Integer, Long> offsets,
+      @QueryParam("resume") @DefaultValue("false") final boolean resume
+  ) throws InterruptedException
+  {
+    if (offsets == null) {
+      return Response.status(Response.Status.BAD_REQUEST)
+                     .entity("Request body must contain a map of { partition:endOffset }")
+                     .build();
+    } else if (!endOffsets.keySet().containsAll(offsets.keySet())) {
+      return Response.status(Response.Status.BAD_REQUEST)
+                     .entity(
+                         String.format(
+                             "Request contains partitions not being handled by this task, my partitions: %s",
+                             endOffsets.keySet()
+                         )
+                     )
+                     .build();
+    }
+
+    for (Map.Entry<Integer, Long> entry : offsets.entrySet()) {
```

Comment:
Hm, while that's true, the only case where nextOffsets would still be incrementing is if the main loop hasn't been paused, and if it hasn't been paused the request will be rejected a few lines later. To be safe, I can move this check after the check to see if the main loop has been paused.